src/main/scala/latis/util/hapi/DataCodec.scala

```diff
@@ -0,0 +1,38 @@
+package latis.util.hapi
+
+import scodec._
+import scodec.bits._
+
+import latis.data.Data
+import latis.data.Data._
+import latis.model._
+
+object DataCodec {
+  private def int2ByteVal(x: Int): ByteValue = ByteValue(x.toByte)
+  private def int2ShortVal(x: Int): ShortValue = ShortValue(x.toShort)
+  private def double2FloatVal(x: Double): FloatValue = FloatValue(x.toFloat)
+
+  def hapiCodec(s: Scalar): Codec[Data] = s.valueType match {
+    case BooleanValueType => codecs.fail(Err("Boolean not supported."))
+    case ByteValueType    => codecs.int32L.xmap[ByteValue](int2ByteVal, _.value.toInt).upcast[Data]
+    case ShortValueType   => codecs.int32L.xmap[ShortValue](int2ShortVal, _.value.toInt).upcast[Data]
+    case IntValueType     => codecs.int32L.xmap[IntValue](IntValue, _.value).upcast[Data]
+    case LongValueType    => codecs.fail(Err("Long not supported."))
+    case FloatValueType   => codecs.doubleL.xmap[FloatValue](double2FloatVal, _.value.toDouble).upcast[Data]
+    case DoubleValueType  => codecs.doubleL.xmap[DoubleValue](DoubleValue, _.value).upcast[Data]
+    case StringValueType if s.metadata.getProperty("size").nonEmpty =>
+      val size = s.metadata.getProperty("size").get.toLong * 8
+      codecs.paddedFixedSizeBits(
+        size,
+        codecs.ascii,
+        codecs.literals.constantBitVectorCodec(BitVector(hex"00"))
+      ).xmap[StringValue](s => StringValue(s.replace("\u0000", "")), _.value).upcast[Data]
+    case StringValueType  => codecs.fail(Err("String without a size defined not supported."))
```

Comment:
I've had the same question about our original binary encoder: How do these failures manifest themselves? When we construct it or when we use it? I'm wondering if we need to fail sooner so we can return a 400 flavor error for an unsupported data type before we start trying to stream data.